drafts/are-you-getting-the-test-pyramid-wrong.md

```diff
@@ -0,0 +1,194 @@
+
+In the `ReportStorageServiceTest`, we've mocked our repository and verified that its `save()` method is called. This is not ideal because we're testing the interactions, not the behaviours. In other words, we're testing that the code calls the `save()` method, but we're not testing that the `save()` method actually persists in the data in the database. This can lead to two problems:
+
+1. The test can pass even if the application code breaks. For example, if the `save()` method throws an exception, the test will still pass because it only verifies that the `save()` method is called.
+2. The test can break even if the application code works. For example, if someone refactors the code to call the `saveAll()` method instead, the test will break because it's expecting the `save()` method to be called.
+
+## What to do instead
+
+Instead of mocking the `RestTemplate` class, why don't we use a fake web server that returns predefined responses? The `MockRestServiceServer` class from the Spring Framework does just that. With much fewer lines of code, we can configure it to return specific responses, without the need for lengthy mocking logic.
+
+In the `ReportStorageServiceTest`, we should care about what matters: is the entity persisted in the repository? We shouldn't get bogged down in how it's persisted, whether it's by using the `repository.save()` or some other method. Verifying the states changed rather than the interactions invoked frees us from testing against the implementation details. This means we don't have to rewrite our tests every time we refactor our code. In this particular example, we can use the `@DataJpaTest` annotation and the `TestEntityManager` class to verify the database state without mocking anything.
+
+## What about speed?
+
+Some developers might argue that avoiding mocks in unit tests makes them slower. But have they become religious about how fast their unit tests should run? Some insist that unit tests must finish within sub-seconds. So far, I've suggested using fakes over mocks in unit tests to simulate behaviours and verify state changes. Apparently, this approach is slower than using mocks and verifying the interactions, but it would still finish within seconds. So let's say it's ten times slower. But does that really mean it is slow, and should we avoid it because of that?
```

Comment:
```suggestion
Some developers might argue that avoiding mocks in unit tests makes them slower. But have they become religious about how fast their unit tests should run? Some insist that unit tests must finish within sub-seconds. So far, I've suggested using fakes over mocks in unit tests to simulate behaviors and verify state changes. Apparently, this approach is slower than using mocks and verifying the interactions, but it would still finish within seconds. So let's say it's ten times slower. But does that really mean it is slow, and should we avoid it because of that?
```